### PR TITLE
Support find nodes via key and add callListener()

### LINKED ISSUE #23: Ability to call listeners from widget harness
**Enhancement**

We should be able to call listeners on a virtual DOM render in a way that is similar to the `widget.sendEvent()` in the harness.  Suggested signature is `widget.callListener('onClick', { /* options? */ })`.

### PR DESCRIPTION
**Pull request by kitsonk**
**Type:** feature

The following has been addressed in the PR:

* [X] There is a related issue
* [X] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [X] Unit or Functional tests are included in the PR

**Description:**

This feature contains several key changes:

- Breaks out the virtual DOM utility functions from `harness` to `support/virtualDom`.  It was getting too confusing and muddled in the `harness` module.  This will be a **breaking change** though, though it should be a simple change import statements in test modules.  I would like to take the hit on this before we end up having to do something later.
- Export two new functions from `support/virtualDom` (one renamed that was previously unexported).  `findKey` finds a virtual DOM node by key and `findIndex` finds a virtual DOM node by index.
- Add `support/callListener` module which has a default export of a function that allows easily calling listeners on virtual DOM.
- Integrate `.callListener()` into `harness`.

These changes have been reflected in the README.

Resolves #23
Refs #9 - lays the groundwork for resolving this PR
Refs #22 - lays the groundwork for resolving this PR

### CODE REVIEW

File: src/harness.ts [hunk 1]
```diff
@@ -284,6 +285,19 @@ export class Harness<P extends WidgetProperties, W extends Constructor<WidgetBas
 	}
 
 	/**
+	 * Call a listener on a target node of the virtual DOM.
+	 * @param method The method to call on the target node
+	 * @param options A map of options that effect the behaviour of `callListener`
```
Comment (dylans) [inline src/harness.ts]: behavior (I think we're trying to be consistent in US English in Dojo 2 docs, though I don't personally care)

File: src/support/callListener.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	 */
+	index?: number | string;
+
+	/**
+	 * Target the node based on a `key`.
+	 */
+	key?: string | object;
+
+	/**
+	 * Supply your own target
```
Comment (dylans) [inline src/support/callListener.ts]: target -> `target`

File: src/support/virtualDom.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+export function assignProperties(target: WNode | HNode, properties: WidgetProperties | VirtualDomProperties): WNode | HNode;
+export function assignProperties(target: WNode | HNode, properties: WidgetProperties | VirtualDomProperties): WNode | HNode {
+	assign(target.properties, properties);
+	return target;
+}
+
+/**
+ * Replace a child of DNode.
+ *
+ * *NOTE:* The replacement modify the passed `target` and does not return a new instance of the `DNode`.
```
Comment (dylans) [inline src/support/virtualDom.ts]: modify -> modifies

File: src/support/virtualDom.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ * @param target The DNode to replace a child element on
+ * @param index A number of the index of a child, or a string with comma seperated indexes that would nagivate
```
Comment (dylans) [inline src/support/virtualDom.ts]: I would say indices is more common than indexes, but I think leave it as-is

File: src/support/virtualDom.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+export function assignProperties(target: WNode | HNode, properties: WidgetProperties | VirtualDomProperties): WNode | HNode {
+	assign(target.properties, properties);
+	return target;
+}
+
+/**
+ * Replace a child of DNode.
+ *
+ * *NOTE:* The replacement modify the passed `target` and does not return a new instance of the `DNode`.
+ * @param target The DNode to replace a child element on
```
Comment (dylans) [inline src/support/virtualDom.ts]: In general I don't like ending phrases in a preposition, but I don't have a better answer here

File: src/support/virtualDom.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ * @param index A number of the index of a child, or a string with comma seperated indexes that would nagivate
+ * @param replacement The DNode to be replaced
+ */
+export function replaceChild(target: WNode | HNode, index: number | string, replacement: DNode): WNode | HNode {
+	/* TODO: Combine with findIndex */
```
Comment (dylans) [inline src/support/virtualDom.ts]: Address TODO or add an issue for this?

File: src/support/virtualDom.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+			if (hasChildren(child)) {
+				return Boolean(found = findKey(child, key));
+			}
+			return false;
+		});
+	return found;
+}
+
+/**
+ * Return a DNode that is identified by supplied index
```
Comment (dylans) [inline src/support/virtualDom.ts]: DNode -> `DNode`

File: src/support/virtualDom.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * @param target The target WNode or HNode to resolve the index for
```
Comment (dylans) [inline src/support/virtualDom.ts]: WNode -> `WNode` and HNode -> `HNode`

**Review by dylans — DISMISSED** [review]
minor nits, code looks reasonable.

File: src/support/virtualDom.ts (hunk 5 above)
Comment (kitsonk) [reply src/support/virtualDom.ts]: http://grammar.yourdictionary.com/parts-of-speech/prepositions/Ending-a-Sentence-with-a-Preposition.html

File: src/support/virtualDom.ts (hunk 4 above)
Comment (kitsonk) [reply src/support/virtualDom.ts]: http://grammarist.com/usage/indexes-indices/

**Review by matt-gadd — APPROVED** [review]

**Pull request merged**